### PR TITLE
[Bug-3140]fix the deadlock between start and stop of ZKServer

### LINKED ISSUE #3140: [BUG] Embed zookeeper server 'ZKServer' has deadlock problem
*For better global communication, please give priority to using English description, thx! *

**Describe the bug**
look at the code snippet
```java
    private static synchronized void startLocalZkServer(final int port, final String dataDirPath,final int tickTime,String maxClientCnxns) {
        if (zkServer != null) {
            throw new RuntimeException("Zookeeper server is already started!");
        }
        zkServer = new PublicZooKeeperServerMain();
        logger.info("Zookeeper data path : {} ", dataDirPath);
        dataDir = dataDirPath;
        final String[] args = new String[]{Integer.toString(port), dataDirPath, Integer.toString(tickTime), maxClientCnxns};

        try {
            logger.info("Zookeeper server started ");
            isStarted.compareAndSet(false, true);

            zkServer.initializeAndRun(args);
        } catch (QuorumPeerConfig.ConfigException e) {
            logger.warn("Caught exception while starting ZK", e);
        } catch (IOException e) {
            logger.warn("Caught exception while starting ZK", e);
        }
    }
```
The above start method will acquire the lock of the ZKServer class. But the line "zkServer.initializeAndRun(args);" will block until the application exit. So it will never release the lock.
```java
    private static synchronized void stopLocalZkServer(final boolean deleteDataDir) {
        // 省略
    }
```
The above stop method try to acquire the lock of the ZKServer class, but it will never succeed.
**To Reproduce**
Steps to reproduce the behavior, for example:

**Expected behavior**

**Screenshots**

**Which version of Dolphin Scheduler:**
 -[1.1.0-preview]

**Additional context**

**Requirement or improvement
- Please describe about your requirements or improvement suggestions.

### PR DESCRIPTION
**Pull request by tswstarplanet**
fix #3140
## *Tips*
- *Thanks very much for contributing to Apache DolphinScheduler.*
- *Please review https://dolphinscheduler.apache.org/en-us/community/index.html before opening a pull request.*

## What is the purpose of the pull request

*(For example: This pull request adds checkstyle plugin.)*

## Brief change log

*(for example:)*
  - Remove the synchronized of start and stop method.
  - Let the ZkServer can start with custom port when execute the "public static void main" method.
  - Refactor the static method to object method to avoid the JMException when execute unit test.

## Verify this pull request

*(Please pick either of the following options)*

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

*(example:)*

  - *Added dolphinscheduler-dao tests for end-to-end.*
  - *Added CronUtilsTest to verify the change.*
  - *Manually verified the change by testing locally.*

### CODE REVIEW

File: dolphinscheduler-service/src/test/java/org/apache/dolphinscheduler/service/zk/ZKServerTest.java [hunk 1]
```diff
@@ -19,17 +19,40 @@
...
+        new Thread(() -> {
+            zkServer.set(new ZKServer());
+            zkServer.get().start();
+        }).start();
+        try {
+            TimeUnit.SECONDS.sleep(5);
+            Assert.assertEquals(true, zkServer.get().isStarted());
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
```
Comment (dailidong) [inline dolphinscheduler-service/src/test/java/org/apache/dolphinscheduler/service/zk/ZKServerTest.java:39]: please use logger.error, not use e.printStackTrace

File: dolphinscheduler-service/src/test/java/org/apache/dolphinscheduler/service/zk/ZKServerTest.java [hunk 2]
```diff
@@ -19,17 +19,40 @@
...
+        AtomicReference<ZKServer> zkServer = new AtomicReference<>();
+        new Thread(() -> {
+            zkServer.set(new ZKServer(2182));
+            zkServer.get().start();
+        }).start();
+        try {
+            TimeUnit.SECONDS.sleep(5);
+            Assert.assertEquals(true, zkServer.get().isStarted());
+        } catch (InterruptedException e) {
+            e.printStackTrace();
```
Comment (dailidong) [inline dolphinscheduler-service/src/test/java/org/apache/dolphinscheduler/service/zk/ZKServerTest.java:54]: the same with the last one

Comment (tswstarplanet) [reply dolphinscheduler-service/src/test/java/org/apache/dolphinscheduler/service/zk/ZKServerTest.java:54]: It's not exactly identical. The last one uses the default port, but this one use a custom port

File: dolphinscheduler-service/src/test/java/org/apache/dolphinscheduler/service/zk/ZKServerTest.java (hunk 1 above)
Comment (tswstarplanet) [reply dolphinscheduler-service/src/test/java/org/apache/dolphinscheduler/service/zk/ZKServerTest.java]: solved

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZKServer.java [hunk 3]
```diff
@@ -34,44 +34,62 @@
...
     /**
      * start service
      */
-    public static void start() {
+    public void start() {
         try {
-            startLocalZkServer(DEFAULT_ZK_TEST_PORT);
+            startLocalZkServer(port);
         } catch (Exception e) {
             logger.error("Failed to start ZK: " + e);
```
Comment (yangyichao-mango) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZKServer.java:88]:
```suggestion
            logger.error("Failed to start ZK ", e);
```

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZKServer.java [hunk 4]
```diff
@@ -94,8 +112,9 @@ public void shutdown() {
      *
      * @param port The port to listen on
      */
-    public static void startLocalZkServer(final int port) {
-        String zkDataDir = System.getProperty("user.dir") +"/zookeeper_data";
+    public void startLocalZkServer(final int port) {
+        String zkDataDir = System.getProperty("user.dir") + ("".equals(prefix) ? prefix : ("/" + prefix)) + "/zookeeper_data";
```
Comment (yangyichao-mango) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZKServer.java:116]:
```suggestion
        String zkDataDir = System.getProperty("user.dir") + (StringUtils.isBlank(prefix) ? StringUtils.EMPTY : ("/" + prefix)) + "/zookeeper_data";
```
Hi,
Good job.
Is it better to use this utils of `org.apache.dolphinscheduler.common.utils.StringUtils`?

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZKServer.java [hunk 5]
```diff
@@ -94,8 +112,9 @@ public void shutdown() {
...
+        logger.warn("The path of zk server exists");
```
Comment (yangyichao-mango) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZKServer.java:117]: Hi,
Why introduce `logger.warn("The path of zk server exists");`?

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZKServer.java [hunk 6]
```diff
@@ -108,11 +127,11 @@ public static void startLocalZkServer(final int port) {
      * @param tickTime    zk tick time
      * @param maxClientCnxns    zk max client connections
      */
-    private static synchronized void startLocalZkServer(final int port, final String dataDirPath,final int tickTime,String maxClientCnxns) {
-        if (zkServer != null) {
+    private void startLocalZkServer(final int port, final String dataDirPath,final int tickTime,String maxClientCnxns) {
+        if (zooKeeperServerMain != null) {
```
Comment (yangyichao-mango) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZKServer.java:131]: Hi,

Is it better to add `isStarted.get()` compare in if statement?

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZKServer.java [hunk 7]
```diff
@@ -121,7 +140,7 @@ private static synchronized void startLocalZkServer(final int port, final String
             logger.info("Zookeeper server started ");
             isStarted.compareAndSet(false, true);
 
-            zkServer.initializeAndRun(args);
+            zooKeeperServerMain.initializeAndRun(args);
         } catch (QuorumPeerConfig.ConfigException e) {
             logger.warn("Caught exception while starting ZK", e);
         } catch (IOException e) {
```
Comment (yangyichao-mango) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZKServer.java:146]:
```suggestion
            zooKeeperServerMain.initializeAndRun(args);
        } catch (QuorumPeerConfig.ConfigException e) {
            logger.warn("Caught exception while starting ZK", e);
        } catch (IOException e) {
```
Hi,

Is it better to merge tow exceptions into one catch block if the two exceptions are handled in the same way?
If it is convenient for you, I think you can modify this part.
Thx a lot for your contribution.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZKServer.java (hunk 5 above)
Comment (tswstarplanet) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZKServer.java:117]: It just warn the previous data may be overwritten.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZKServer.java (hunk 7 above)
Comment (tswstarplanet) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZKServer.java]: solved

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZKServer.java (hunk 6 above)
Comment (tswstarplanet) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZKServer.java]: solved

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZKServer.java (hunk 4 above)
Comment (tswstarplanet) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZKServer.java]: solved

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZKServer.java [hunk 8]
```diff
@@ -108,31 +131,29 @@ public static void startLocalZkServer(final int port) {
...
      */
-    private static synchronized void startLocalZkServer(final int port, final String dataDirPath,final int tickTime,String maxClientCnxns) {
-        if (zkServer != null) {
-            throw new RuntimeException("Zookeeper server is already started!");
-        }
-        zkServer = new PublicZooKeeperServerMain();
-        logger.info("Zookeeper data path : {} ", dataDirPath);
-        dataDir = dataDirPath;
-        final String[] args = new String[]{Integer.toString(port), dataDirPath, Integer.toString(tickTime), maxClientCnxns};
+    private void startLocalZkServer(final int port, final String dataDirPath,final int tickTime,String maxClientCnxns) {
```
Comment (yangyichao-mango) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZKServer.java:134]:
```suggestion
    private void startLocalZkServer(final int port, final String dataDirPath, final int tickTime, String maxClientCnxns) {
```
Hi,
Thx a lot for your good job and contribution.
Please configure the `checkstyle.xml` plugin[1] to your IDE to format code.
[1] https://github.com/apache/incubator-dolphinscheduler/tree/dev/style

**Pull request merged**